src/jdk.compiler/share/classes/com/sun/tools/javac/util/Convert.java

```diff
@@ -223,10 +223,11 @@ public static String utf2string(byte[] src, int sindex, int len, Validation vali
      */
     public static int utfNumChars(byte[] buf, int off, int len) {
         int numChars = 0;
-        while (len-- > 0) {
-            int byte1 = buf[off++];
-            if (byte1 < 0)
-                len -= ((byte1 & 0xe0) == 0xc0) ? 1 : 2;
+        while (len > 0) {
+            int byte1 = buf[off];
+            int nbytes = byte1 >= 0 ? 1 : (byte1 & 0xe0) == 0xc0 ? 2 : 3;
+            off += nbytes;
+            len -= nbytes;
             numChars++;
```

Comment:
I wonder if it wouldn't be easier to simply ignore bytes from `buf` in the form of `0b10xxxxxx`? E.g. something along these lines:
```suggestion
            int byte1 = buf[off++];
            if ((byte1 & 0b11000000) != 0b10000000) {
                //only count the first byte in every encoded sequence,
                //and ignore the other:
                numChars++;
            }
```

Reply:
Nice! Thanks, fixed in 91a9037f434. FYI I used `0x` instead of `0b` to stay consistent with the rest of the class.